### PR TITLE
Add identity_conversion lint

### LINKED ISSUE #1051: Lint "useless" `into` conversions
Lint uses of .into() where the result type is the same as the input type.
# Example

``` Rust
#[derive(Debug)]
struct Foo {
    // Used to be `String`, so calling `.into()` on a literal made sense.
    // It got refactored to be `&str`, so calling `.into()` doesn't make sense anymore.
    name: &'static str,
}

fn main() {
    let f = Foo { name: "Hello".into() };
    println!("{:?}", f);
}

```

### PR DESCRIPTION
**Pull request by sinkuu**
Fixes #1051.

### CODE REVIEW

File: clippy_lints/src/identity_conversion.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        body: &'tcx Body,
+        _: Span,
+        _: NodeId
+    ) {
+        let mut vis = IdentityConversionVisitor { cx: cx };
+        walk_expr(&mut vis, &body.value);
+    }
+}
+
+/// Visiting expressions except ones expanded from `?` (that contains `From::from`)
```
Comment (oli-obk) [inline clippy_lints/src/identity_conversion.rs]: couldn't you use `check_expr` and add an `in_macro` check?

Using a custom visitor breaks fine grained `#[allow]` attributes

Comment (sinkuu) [inline clippy_lints/src/identity_conversion.rs]: It didn't work. It seems that expressions expanded from `?` don't have expn_info except `Carrier::translate`/`Carrier::from_error` call.

Comment (oli-obk) [inline clippy_lints/src/identity_conversion.rs]: that sounds like a bug in rustc: https://github.com/rust-lang/rust/blob/master/src/librustc/hir/lowering.rs#L1997 Let's fix it there first.

Comment (sinkuu) [inline clippy_lints/src/identity_conversion.rs]: Adding expn_info made compiler put `in this macro expansion` here and there in error, which seems a regression to me. I'll try to suppress them.

**Comment by sinkuu** [comment]
Sorry for being late. It's working without custom visitor now. Can anyone re-review this?

**Comment by oli-obk** [comment]
Awesome! Thanks for coming back to this, lgtm

**Pull request merged**